File: src/foam/swift/parse/json/KeyParser.js
/**
 * @license
 * Copyright 2017 The FOAM Authors. All Rights Reserved.
 * http://www.apache.org/licenses/LICENSE-2.0
 */

foam.CLASS({
  package: 'foam.swift.parse.json',
  name: 'KeyParser',
  extends: 'foam.swift.parse.parser.ProxyParser',
  requires: [
    'foam.swift.parse.parser.Alt',
    'foam.swift.parse.parser.Literal',
  ],
  axioms: [
    foam.pattern.Multiton.create({ property: 'key' })
  ],
  properties: [
    {
      class: 'String',
      name: 'key',
    },
    {
      name: 'delegate',
      swiftExpressionArgs: ['key'],
      swiftExpression: function() {/*
return self.Alt_create(["parsers": [
  self.Literal_create(["string": "\"" + key + "\""]),
  self.Literal_create(["string": key]),
]])
      */},
    },
  ],
});

File: src/foam/lib/json/KeyParser.java
/**
 * @license Copyright 2017 The FOAM Authors. All Rights Reserved.
 * http://www.apache.org/licenses/LICENSE-2.0
 */

package foam.lib.json;

import foam.lib.parse.*;

public class KeyParser
  implements Parser
{
  protected Parser delegate;

  public KeyParser(String key) {
    delegate = new Alt(
      Literal.create("\"" + key + "\""),
      Literal.create(key));
  }

  public PStream parse(PStream ps, ParserContext x) {
    return ps.apply(delegate, x);
  }
}
